### ISSUE TITLE
"auto" layout algorithm should be able to generate any layout

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#4956]
Currently the [`"auto"` layout algorithm](https://gpuweb.github.io/gpuweb/#default-pipeline-layout) makes a few assumptions about what layout to generate, which means if those assumptions are wrong for your application, you're forced to use a explicit layout. At least the following aren't achievable - not sure if I'm missing any:

- `"non-filtering"` samplers (would be partially but not entirely fixed by #4952)
  - and `"unfilterable-float"` bindings to go with them

Also the following, but I believe they don't matter because of auto layout exclusivity:

- `"float"` when there are no usages with a sampler (it picks `"unfilterable-float"`)
- `"depth"` when the shader type is non-depth (`texture_2d<f32>` etc.)

**Comment by kainino0x** [comment]
A simple proposal for 1 and 2 is to allow `@filter(true)` and `@filter(false)` on both sampler and texture bindings which choose between `"filtering"`/`"float"` and `"non-filtering"`/`"unfilterable-float"`.

For 3, something like `texture_2d<depth>` or `@depth texture2d<f32>`?

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2024‐11‐19-Minutes#proposal-auto-layout-algorithm-should-be-able-to-generate-any-layout--issue-4956">WGSL 2024-11-19 Minutes</a></summary>

* KG: We didn’t have much time to go over this, so we need more time to give it more thought.
* ds: Some things we need, like unfilterable-floats, so if we give a way to say so, we wouldn’t have to guess in our implementation.
* **(no resolution today)**


</details>

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2024‐12‐03-Minutes#proposal-auto-layout-algorithm-should-be-able-to-generate-any-layout--issue-4956">WGSL 2024-12-03 Minutes</a></summary>

*  JB: Does it make sense that attributes apply to the global var and not the wgsl type of the sampler? Is this a thing that everyone lets us do whatever we want and we can just leave it polymorphic, or do we want to make it part of the type system?
* DN: Good question. :) 
* **(Out of time)**

</details>

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2024‐12‐10-Minutes#proposal-auto-layout-algorithm-should-be-able-to-generate-any-layout--issue-4956">WGSL 2024-12-10 Minutes</a></summary>

* **KG: **Do we think we should wait for a proposal, like the similar other one?
* DN: I think there’s going to be a trend of having less strict types in the shader, and this kind of goes the other way instead. E.g. types of textures like rgba8.
* Ds: But this just gives auto the ability to work correctly for cases where our current more naive model wouldn’t be able to.
* DN: This looks like it should be type information, but we’re adding it as an attribute instead.
* DN: I’d rather have the convenience pay for itself on its own.
* Ds: An annotation feels cheaper to add, but we could actually make types for this as well if we wanted to.
* AB: So is auto like, the main feature that we build things around, or should we accept that it’s incomplete? It sounds like we’re saying that auto is so key that we want to ensure that we can always use it.
* Ds: Unclear, kind of a question for the API-side people?
* JB: [...] The only reason you would need to not use auto, is that you can share them. What I’m hoping to get from this is to never had to write types in JS again, given that I’ve already written the types in WGSL.
* KG: I think what you said stands on this issue. We’ve heard from people the idea of “why do i have to write these types; the types are already in WGSL”. There is a desire to allow people to use that mode of operation without ever having to do explicit bind group typing. They are really nice-to-have things that don’t seem terribly difficult to add support; it’s a matter of figuring out how we’re going to do it, hence type vs. annotation.
* KG: Think **next step is to have a proposal to discuss**.  Don’t think I heard we shouldn’t support the use case.

</details>

**Comment by kainino0x** [comment]
I think the only thing that's really necessary right now (pre-#4957) is the ability to say a sampler is non-filtering. If we can do that then we can update the layout algorithm so it generates unfilterable-float texture bindings if the texture is only used with non-filtering samplers.

I do not think it should be part of the type because then it will percolate through the type system - especially not ideal if we can pass pointers to sampler bindings as function arguments. (I forget if this is possible now but it's one of the proposals for #4957). It should stay only at the interface, because only the interface cares about it. I am interested if WGSL experts agree on this.

Finally it would make sense to have something that is consistent for all of the cases we would want to solve for #4957 (listed in OP). For that reason I would initially propose something matching the API side exactly:
`@group(0) @binding(0) @sampler_binding_type(filtering) var mySampler: sampler;`
`@group(0) @binding(0) @sampler_binding_type(non_filtering) var mySampler: sampler;`
extendable to:
`@group(0) @binding(1) @texture_sample_type(float) var myTexture: texture_2d<f32>;`
`@group(0) @binding(1) @texture_sample_type(unfilterable_float) var myTexture: texture_2d<f32>;`
`@group(0) @binding(1) @texture_sample_type(depth) var myTexture: texture_2d<f32>;`

Beyond that it's just bikeshedding. (The names do not have to be this verbose.)

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2025‐02‐18-Minutes#proposal-auto-layout-algorithm-should-be-able-to-generate-any-layout--issue-4956">WGSL 2025-02-18 Minutes</a></summary>

* KG: Last comment from Kai, last thing we need is a way for a sampler to say non-filtering. Would allow auto-layout generated with the  needed things. Theoretically possible to have the us determine by use but if we did that it would be spooky action at a distance as you change how you access the texture and bindings would fail to attach. This moves the spooky action to less far away conceptually. It's written into the binding if you expect it to be filterable or not. Possible we're just waiting for a proposal
* MW: Was wondering if we considered how this may interact with the bindless proposal which is on-going. If that's going to impact how useful this potentially is.
* KG: Good question i don't know
* DN: In general, how does bindless interact with auto-layout at all? (I don't know)
* MW: Yo don't know your bindings at shader compilation time but you may not know your layout until the encoding time.
* JB: There was discussion in previous meetings about if these annotations (attributes) really belonged in the sampler type and i didn't see in the minutes a clear decision aboutthat. Seemed like DN wanted to think about it.
* DN: If you encode this into the type you have this additional rigidity where you can't pass the samplers because we dont' have templated parameter types which seems annoying and a lot of hoops for no reason. So we'd be  modifying a bunch of shader code for things which aren't a shader concern. It's a n API concern. Better to have as an attribute on the var and not the type.
* JB: Sounds great
* DN: If we had union types but we don't and won't any time soon
* JB: Or generics or something.
* **KG: Sounds like waiting on proposal. Kai had a tiny sketch of a proposal and sounds like if made into PR we'd consider it.**

</details>
